### PR TITLE
Add recovery state tracking for Searchable Snapshots.

### PR DESCRIPTION
**Pull request by fcofdez**
This pull request adds recovery state tracking for Searchable Snapshots.

In order to track recoveries for searchable snapshot backed indices, this pull
request adds a new setting to the indices, index.recovery.type. This allows to
inject a RecoveryState instance that's able to deal with the small differences
that arise during index recoveries backed by a snapshot.

Those differences can be summarized as follows:
- The Directory implementation that's provided by SearchableSnapshots mark the
    snapshot files as reused during recovery. In order to keep track of the 
    recovery process as the cache is filled, those files shouldn't be marked
    as reused.
- In some cases the cache can start a pre-warming phase before the files that
    would be recovered are known to `RecoveryState`. To deal with that fact, the
    files that are downloaded during the pre-warming phase are tracked in a
    temporal `PersistentCacheTracker`. Once the shard is started, a
    `OnDemandRecoveryState` is injected into the searchable snapshot directory,
    and the information tracked during the pre-warming phase is transferred to
    the shard `RecoveryState`.

Finally, it introduces a new final stage, `ON_DEMAND`, for searchable snapshot
recoveries. This prevents some confusion when the stage is `DONE` but the stats
keep changing.

### CODE REVIEW

**Comment by fcofdez** [comment]
jenkins retest this please

**Comment by fcofdez** [comment]
@elasticmachine run elasticsearch-ci/2

**Comment by fcofdez** [comment]
@elasticmachine test this

**Comment by fcofdez** [comment]
jenkins retest this please

**Comment by fcofdez** [comment]
@elasticmachine run elasticsearch-ci/2

**Comment by tlrx** [comment]
I skimmed over the changes and this looks like the right direction. Maybe we could split this into smaller pull requests that would make the review easier?

**Comment by fcofdez** [comment]
Thanks for the feedback @tlrx!. Sure, I'll split it into a couple of PRs 👍 .

**Comment by fcofdez** [comment]
I've divided this PR into:
- https://github.com/elastic/elasticsearch/pull/59038
- https://github.com/elastic/elasticsearch/pull/59039
- A third PR with the Recovery stats for searchable snapshots once we merge the former PRs

File: server/src/main/java/org/elasticsearch/indices/recovery/RecoveryState.java
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/indices/recovery/RecoveryState.java:644]: I'm puzzled by this: we should know the lengths of all the files in the snapshot before we start doing any meaningful work in the recovery. Can you explain more clearly why we need to account for files of unknown length?

**Review by DaveCTurner — COMMENTED** [review]
I left some comments on the subsidiary PRs, but added some other ones here too.

File: server/src/main/java/org/elasticsearch/indices/recovery/RecoveryState.java
Comment (fcofdez) [reply server/src/main/java/org/elasticsearch/indices/recovery/RecoveryState.java:644]: It's possible that the cache starts pre-warming before the file details are known (as we load the snapshot data and prewarm the cache before the shard has recovered) , hence the reason to deal with files with unknown length.

File: server/src/main/java/org/elasticsearch/index/IndexService.java
Comment (fcofdez) [reply server/src/main/java/org/elasticsearch/index/IndexService.java:155]: I fixed that on the other PRs.

File: server/src/main/java/org/elasticsearch/indices/recovery/RecoveryState.java
Comment (fcofdez) [reply server/src/main/java/org/elasticsearch/indices/recovery/RecoveryState.java:81]: I agree, I don't like that stage name that much, thanks for the suggestions.

File: server/src/main/java/org/elasticsearch/indices/recovery/RecoveryState.java
Comment (fcofdez) [reply server/src/main/java/org/elasticsearch/indices/recovery/RecoveryState.java:234]: I wasn't 100% sure about the semantics we want here. The shard is "recovered" at that point but we still modify the recovery data as we download/evict data on the cache, so the recovery for the index is still active. In any case, we keep the timer for the entire recovery running too, so maybe it's enough with that, wdyt?

Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/indices/recovery/RecoveryState.java:234]: On reflection this is more tricky; if we claim the index phase _is_ ended then it seems confusing to keep on updating its stats. Can we avoid stopping the index timer entirely?

File: server/src/main/java/org/elasticsearch/indices/recovery/RecoveryState.java
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/indices/recovery/RecoveryState.java:644]: Yes, this seems wrong, by the time we start pre-warming we definitely know the sizes of all the files, although we may not have exposed them to the recovery state yet. Put differently, it seems problematic that we are updating the index-phase recovery stats before we enter the index phase of the recovery. Can we defer those stats updates somehow?

Comment (fcofdez) [reply server/src/main/java/org/elasticsearch/indices/recovery/RecoveryState.java:644]: I thought about triggering the pre-warming phase just after the shard has started, that would solve the issue. But I'm not sure if we want to start pre-warming the cache as soon as possible to minimize the initial search latency. An alternative could be to inspect the `IndexInputStats` once the shard has started and update the `RecoveryState` accordingly.  I think the later approach would simplify things and would avoid to have to keep track of files with unknown length.

File: server/src/main/java/org/elasticsearch/indices/recovery/RecoveryState.java
Comment (fcofdez) [reply server/src/main/java/org/elasticsearch/indices/recovery/RecoveryState.java:234]: yes, we can intercept the call to this method on the `LazyRecoveryState` and avoid stopping the index timer for that state transition.

**Comment by fcofdez** [comment]
There are still some tests missing (mostly around cache evictions), but I would like to get another review to check if the new approach makes sense.

File: server/src/main/java/org/elasticsearch/indices/recovery/RecoveryState.java
Comment (tlrx) [reply server/src/main/java/org/elasticsearch/indices/recovery/RecoveryState.java]: Another option is ` CACHED`, which I find more explicit  even more explicit

**Comment by tlrx** [comment]
Thanks @fcofdez. I looked at the code and I'm wondering if we could make things a bit more simpler. 

My understanding is that in the current code recovery files details are added as usual during recovery, and in the case of prewarming (which is enabled by default) the file details are tracked but merged alltogether when the shard is STARTED. It means that the shard does not report the exact state of what is on disk/cache until the shard reached a given state. We expect the shards to be started quickly but I think this can be confusing to other parts of the system. I also think it introduces a bit more complexity in the code.

My suggestion would be to initialize the recovery files details at the same time the files are exposed by the Lucene directory. This would happen in the pre recovery hook before the recovery kicks in. This way we would have a coherency between the files exposed by the directory and the recovery files details, and the files details would exist before the prewarming kicks in too. The directory could then update the # cached bytes written as usual and that would update both the index input stats and the recovery files details. That looks simpler to me and brings the benefit of reflecting what is exactly on cache at the very beginning of the recovery. @fcofdez noticed that reporting file changes while the recovery is still in `INDEX` or `TRANSLOG` stages can be confusing. My opinion is that we expect those stages to be executed almost immediately for searchable snapshot shards so it should be OK as those shards have by nature a different recovery process as regular shards.

@DaveCTurner we'd be happy to have your opinion on this suggestion ^

**Comment by fcofdez** [comment]
Thanks for the review @tlrx, finally I went ahead and made the changes that you suggested. As you stated, this new solution is simpler.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshotsConstants.java
Comment (tlrx) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshotsConstants.java:33]: I think ` on_demand` would be more coherent with the corresponding class name

**Review by tlrx — COMMENTED** [review]
This looks good - I've left a first bunch of comments and I'll look deeper at the eviction listener / resetting recovery file detail.

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/recoveries/OnDemandRecoveryState.java
Comment (fcofdez) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/recoveries/OnDemandRecoveryState.java:86]: During Peer Recovery the file details are cleared if the process needs to wait until the shard is allocated, so we need to keep track of which file details came from the snapshot and keep those around.

Comment (tlrx) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/recoveries/OnDemandRecoveryState.java:86]: I'd expect no files to be recovered from peer at all, what are those files?

Comment (fcofdez) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/recoveries/OnDemandRecoveryState.java:86]: I think this was a misunderstanding on my side, somehow I thought that a new `segments_n` file was added during peer recovery, but it doesn't seem the case. I'll remove this.

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotDirectory.java
Comment (tlrx) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotDirectory.java:378]: Registering another EvictionListener looks trappy to me as we can't really be sure that the CacheFile instance returned here will be the same as the one used in CachedBlobContainerIndexInput.readInternal().

I'm wondering if we could instead push down the logic of resetting the recovery file details into CacheFileReference ? So that every time a new CacheFileReference is acquired we reset the recovery details? WDYT?

Comment (fcofdez) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotDirectory.java:378]: I tried that approach, the problem was that since we release the reference to the `CacheFile` once we close `CachedBlobContainerIndexInput`, it's possible that when we clear the cache the eviction listener on `CacheFileReference` is not triggered as the reference was released already. Alternatively what we could do is reseting the recovered bytes in `SearchableSnapshotDirectory#clearCache` and push down the eviction logic to `CacheFileReference`. WDYT?

Comment (tlrx) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotDirectory.java:378]: Yes, I think it is worth trying this

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/recoveries/OnDemandRecoveryState.java
Comment (fcofdez) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/recoveries/OnDemandRecoveryState.java:68]: I think that those methods are synchronized so the reporting that `RecoveryState` produces is consistent across the different related values (i.e. `recoveredFileCount()` and `recoveredFilePercent()`?  Maybe we could make a copy for the reporting methods (i.e. `toXContent`) so those can operate over a consistent view?

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CachedBlobContainerIndexInput.java
Comment (fcofdez) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CachedBlobContainerIndexInput.java:526]: >I wish we could avoid doing something like this, because adding eviction listeners has an impact on file handles and we don't want to keep a FileChannel opened for the cache file if there's no more IndexInput opened on it.

It's true that with the current approach it's possible that the reference for the `CacheFile` can live longer than needed, i.e. all `CachedBlobContainerIndexInput` are closed but we still have room in the cache and a `RecoveryStateCacheFileEvictionListener` holds a reference.

> I think that this RecoveryStateCacheFileEvictionListener is just needed in case of IndexInput being all close and CacheFile not evicted yet, so that when it will be evicted we can still update the recovery file details for it?

That's correct. When we `IndexInput` is closed we don't know if there are other references pointing to the same underlying `CacheFile` so we cannot remove it from the `FileDetails`.

> I'm wondering if we could keep the CacheFile in the CachedFileDetail and check in the recovered() method if it's being evicted, and then remove it from the cacheFiles Set there?

There's a minor problem with that approach, if the `FileInfo#recovered()` is not consumed those `CacheFile` instances could live forever, but I guess that should be ok? I'm not sure how common is to have multiple `IndexInput` pointing to the same `CacheFile` triggering this scenario.

Comment (tlrx) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CachedBlobContainerIndexInput.java:526]: Thanks for your response, you perfectly understood what I might despite the poor quality of my previous comment.



> There's a minor problem with that approach, if the `FileInfo#recovered()` is not consumed those `CacheFile` instances could live forever, but I guess that should be ok?

I thought about it too, and I think we can circumvent that by also cleaning up the `CacheFile` set every time a CacheFile is added/removed or recovered bytes computed?

File: x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshotsIntegTests.java
Comment (fcofdez) [reply x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshotsIntegTests.java:605]: Do you mean that we should check that the number of shards match the number of `RecoveryState` we get back?

Comment (tlrx) [reply x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshotsIntegTests.java:605]: Yes

**Comment by fcofdez** [comment]
@elasticmachine update branch

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotDirectory.java
Comment (tlrx) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotDirectory.java:145]: Can be removed

File: x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshotsIntegTests.java
Comment (tlrx) [inline x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshotsIntegTests.java:740]: There's already a `getNumShards()` method available for this - it rreturn the number of primaries, replicas and data copies (basically what you did here)

File: x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshotsRecoveryStateTests.java
Comment (tlrx) [inline x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshotsRecoveryStateTests.java:81]: Can we also check the values before clearing the cache (so that we ensure clear cache had an action)?

File: x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshotsRecoveryStateTests.java
Comment (tlrx) [inline x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshotsRecoveryStateTests.java:74]: there's no official guidance for this, but we tend to use `equalTo()` even for primitive types (except maybe `boolean`)

**Review by tlrx — COMMENTED** [review]
Nice work @fcofdez - I've noticed things that need to be addressed and we should be good after that

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/SparseFileTracker.java
Comment (fcofdez) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/SparseFileTracker.java:51]: As you point it's a tricky situation. I added this method since the file systems that support `SPARSE` files, report only the written data. Meaning that during testing we check the physical length of files at disk and when we compare it with the `RecoveryState` information we get a mismatch if we use the file length. Maybe we should just rely on the OS to report the cache file physical size using the file descriptor available at `CacheFile`? wdyt @tlrx

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CacheFile.java
Comment (tlrx) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CacheFile.java:83]: nit: can we add some doc?

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/CacheService.java
Comment (tlrx) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/CacheService.java:54]: Can you revert this?

**Review by tlrx — APPROVED** [review]
LGTM, thanks @fcofdez ! I left very minor comments

**Comment by fcofdez** [comment]
@elasticmachine update branch

**Comment by fcofdez** [comment]
I've opened a new PR #60505 with the changes discussed offline. I'll close this one.

**Pull request closed without merging**
